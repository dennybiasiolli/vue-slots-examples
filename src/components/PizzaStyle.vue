<script setup lang="ts">
// credits to https://codepen.io/baumannzone/pen/ZEBYEOm
import { CrustComposition, CrustStyle, Sauce, Topping } from '@/types'
import { computed } from 'vue'

export interface Props {
  crustStyle: {
    key: CrustStyle
    name: string
    size: string
  }
  crustComposition: {
    key: CrustComposition
    name: string
    color: string
  }
  sauce: {
    key: Sauce
    name: string
    color: string
  }
  toppings: {
    key: Topping
    name: string
  }[]
}

const props = defineProps<Props>()

const chorreoColor = computed(() => {
  if (props.toppings.some((topping) => topping.key===Topping.MOZZARELLA)) {
    return "#ebe5b2";
  }
  return props.sauce.color;
})
</script>

<template>
  <div class="pizza-container">
    <div class="pizza">
      <div
        v-for="topping in toppings"
        :key="topping.key"
        class="topping"
        :class="topping.key"
      >
        <div v-if="topping.key === 'basil'" class="leaf"></div>
      </div>
      <div class="slices">
        <div class="slice slice-1"></div>
        <div class="slice slice-2"></div>
        <div class="slice slice-3"></div>
        <div class="slice slice-4"></div>
      </div>
    </div>
  </div>
</template>

<style scoped>
.pizza-container {
  display: flex;
  justify-content: center;
  margin: 20px;
}

.pizza {
  position: relative;
  width: 400px;
  height: 400px;
  background-color: v-bind('sauce.color');
  border-radius: 50%;
  box-shadow: 0 0 0px v-bind('crustStyle.size') v-bind('crustComposition.color');
}

.pizza:after {
  content: '';
  position: absolute;
  width: 10px;
  height: 50px;
  left: 198px;
  background-color: v-bind('chorreoColor');
  border-radius: 0 0 10px 10px;
  box-shadow: 12px -1px v-bind('chorreoColor');
  transform: translateX(3px) translateY(-1px);
  animation: chorreo 4s ease-out infinite;
}

.pizza .topping {
  position: absolute;
  width: 400px;
  height: 400px;
  border-radius: 50%;
  overflow: hidden;
}

.pizza .mozzarella {
  z-index: 0;
  background-image: radial-gradient(circle at 30px 30px, #ebe5b2 5px, transparent 60px);
  background-size: 60px 60px;
}

.pizza .basil .leaf {
  position: absolute;
  width: 25px;
  height: 25px;
  background-color: #4caf50;
  z-index: 13;
  border-radius: 100px 0;
  box-shadow: 50px 100px #4caf50, 150px 120px #4caf50, 130px 40px #4caf50, 200px 250px #4caf50, 250px 300px #4caf50, 300px 100px #4caf50, 350px 200px #4caf50, 35px 280px #4caf50, 150px 320px #4caf50, 130px 40px #4caf50, 100px 200px #4caf50, 250px 60px #4caf50, 230px 160px #4caf50;
}

.pizza .pepperoni {
  transform: rotate(15deg);
  z-index: 1;
  background-image: radial-gradient(
    circle at 40px 40px,
    #ea5353 14px,
    #e21c1c 16px,
    transparent 18px
  );
  background-size: 80px 80px;
}

.pizza .onions {
  z-index: 2;
  background-image: radial-gradient(
      circle at 10px 0px,
      transparent 6px,
      #d2d2d2 5px,
      #f5f5f5 11px,
      transparent 13px
    ),
    radial-gradient(
      circle at 0px 10px,
      transparent 6px,
      #d2d2d2 5px,
      #f5f5f5 11px,
      transparent 13px
    );
  background-size:
    80px 90px,
    150px 150px;
}

.pizza .olives {
  z-index: 3;
  transform: rotate(25deg);
  background-image: radial-gradient(
    circle at 70px 70px,
    #455a64 3px,
    #374850 9px,
    transparent 10px
  );
  background-size: 100px 100px;
}

.pizza .egg {
  transform: rotate(15deg);
  z-index: 1;
  background-image: radial-gradient(
    circle at 40px 40px,
    #f57f18 5px,
    #ffefc1 16px,
    transparent 18px
  );
  background-size: 100px 100px;
}

.pizza .zucchini {
  transform: rotate(45deg);
  z-index: 1;
  background-image: radial-gradient(
    circle at 40px 40px,
    #2dba20 14px,
    #0d4d07 16px,
    transparent 18px
  );
  background-size: 60px 60px;
}

.pizza .cherry-tomatoes {
  transform: rotate(70deg);
  z-index: 3;
  background-image: radial-gradient(
    circle at 30px 30px,
    #f52525 14px,
    #f52525 16px,
    transparent 18px
  );
  background-size: 80px 80px;
}

.pizza .slices {
  display: flex;
  position: relative;
  width: 100%;
  height: 100%;
  justify-content: center;
}

.pizza .slices .slice {
  width: 1px;
  height: 100%;
  background-color: v-bind('crustComposition.color');
}

.pizza .slices .slice-1 {
  transform: rotate(45deg);
}

.pizza .slices .slice-2 {
  transform: rotate(90deg);
}

.pizza .slices .slice-3 {
  transform: rotate(135deg);
}

.pizza .slices .slice-4 {
  transform: rotate(180deg);
}

@-moz-keyframes chorreo {
  50% {
    height: 100px;
  }
}
@-webkit-keyframes chorreo {
  50% {
    height: 100px;
  }
}
@-o-keyframes chorreo {
  50% {
    height: 100px;
  }
}
@keyframes chorreo {
  50% {
    height: 100px;
  }
}
</style>
