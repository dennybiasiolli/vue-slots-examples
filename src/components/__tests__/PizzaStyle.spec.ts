import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import PizzaStyle from '../PizzaStyle.vue'
import {
  CrustComposition,
  CrustStyle,
  Sauce,
  Topping,
  crustCompositionMap,
  crustStyleMap,
  sauceMap,
  toppingMap
} from '@/types'

describe('PizzaStyle', () => {
  it('without pineapple', () => {
    const wrapper = mount(PizzaStyle, {
      props: {
        crustStyle: crustStyleMap[CrustStyle.PAN],
        crustComposition: crustCompositionMap[CrustComposition.WHOLE],
        sauce: sauceMap[Sauce.TOMATO],
        toppings: [toppingMap[Topping.MOZZARELLA], toppingMap[Topping.BASIL]]
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('img').exists()).toBe(false)
  })

  it('with pineapple', () => {
    const wrapper = mount(PizzaStyle, {
      props: {
        crustStyle: crustStyleMap[CrustStyle.CRISPY],
        crustComposition: crustCompositionMap[CrustComposition.WHITE],
        sauce: sauceMap[Sauce.OIL],
        toppings: [toppingMap[Topping.PEPPERONI], toppingMap[Topping.PINEAPPLES]]
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('img').exists()).toBe(true)
  })
})
