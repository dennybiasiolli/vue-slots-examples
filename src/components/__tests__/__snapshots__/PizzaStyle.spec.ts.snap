// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PizzaStyle > with pineapple 1`] = `
"<img data-v-2d497697=\\"\\" class=\\"massacred-meme\\" alt=\\"massacred meme\\" src=\\"/src/assets/massacred-meme.png\\">
<div data-v-2d497697=\\"\\" class=\\"pizza-container\\">
  <div data-v-2d497697=\\"\\" class=\\"pizza\\">
    <div data-v-2d497697=\\"\\" class=\\"topping pepperoni\\">
      <!--v-if-->
    </div>
    <div data-v-2d497697=\\"\\" class=\\"topping pineapples\\">
      <!--v-if-->
    </div>
    <div class=\\"slices\\" data-v-2d497697=\\"\\">
      <div class=\\"slice slice-1\\" data-v-2d497697=\\"\\"></div>
      <div class=\\"slice slice-2\\" data-v-2d497697=\\"\\"></div>
      <div class=\\"slice slice-3\\" data-v-2d497697=\\"\\"></div>
      <div class=\\"slice slice-4\\" data-v-2d497697=\\"\\"></div>
    </div>
  </div>
</div>"
`;

exports[`PizzaStyle > without pineapple 1`] = `
"<!--v-if-->
<div data-v-2d497697=\\"\\" class=\\"pizza-container\\">
  <div data-v-2d497697=\\"\\" class=\\"pizza\\">
    <div data-v-2d497697=\\"\\" class=\\"topping mozzarella\\">
      <!--v-if-->
    </div>
    <div data-v-2d497697=\\"\\" class=\\"topping basil\\">
      <div data-v-2d497697=\\"\\" class=\\"leaf\\"></div>
    </div>
    <div class=\\"slices\\" data-v-2d497697=\\"\\">
      <div class=\\"slice slice-1\\" data-v-2d497697=\\"\\"></div>
      <div class=\\"slice slice-2\\" data-v-2d497697=\\"\\"></div>
      <div class=\\"slice slice-3\\" data-v-2d497697=\\"\\"></div>
      <div class=\\"slice slice-4\\" data-v-2d497697=\\"\\"></div>
    </div>
  </div>
</div>"
`;
